### PR TITLE
RobotPoseEstimator

### LINKED ISSUE #522: PhotonLib - Add observation fusion strategies
After https://github.com/PhotonVision/photonvision/pull/521 is merged....

Expand PhotonLib to have utility classes to help users merge the two reported poses for each target to avoid ambiguity issues.

Strategies could support:

* Multiple cameras looking at the same target, and needing to pick the pose which is most common between them
* One camera looking at multiple targets, again needing to pick the most common pose answer
* A camera seeing one target, with an assumed height off the ground, picking the pose which places it to the assumed height
* A camera seeing one target, and picking a pose most similar to the most recently observed pose
* A camera seeing one target, and picking a pose most similar to one provided externally (ie, from previous loop's odometry)
* A camera seeing one target, and picking the pose with the lowest ambiguity.

I'm not entirely certain exactly how to architect this, just stating some requirements to get the ball rolling.

### PR DESCRIPTION
**Pull request by realSaddy**
Closes #522 
WIP

Strategies:
| Strategy	| Java 	| C++| 	
|---	|---	|---	|
|   LOWEST_AMBIGUITY	|   Done |   Done	|
|   CLOSEST_TO_CAMERA_HEIGHT	|   Done |   Done	|
|   CLOSEST_TO_REFERENCE_POSE	|   Done	|   Done	|
|   CLOSEST_TO_LAST_POSE |   Done |   Done	| 
|   AVERAGE_BEST_TARGETS |   Done |   Done	|
|   TRILERATION	|   Outside scope |   Outside scope 	|

Etc.
- **Reject impossible** - Outside scope

### CODE REVIEW

**Comment by gerth2** [comment]
I like the start so far.

General thought - since we're supporting multiple cameras...

In theory, multiple cameras seeing the same target (or multiple targets from one camera) will produce a spread of pose estimates. Half should be roughly-clustered together, the other half will be spread out and "off in the weeds". There could be a strategy which rejects the outliers and just averages the cluster (perhaps weighted average by ambiguity?)

Other thoughts on having configurable rejection strategies for poses that are off the field, or floating in mid air?

**Comment by NoahStricker** [comment]
Our tests concluded that the rotation output from PNP was a lot less accurate than the distance from the tags. How about using trilateration on the 2 best tags ignoring rotation? Another plausible method is to run PNP with all the tags together and use the estimate from that.

**Comment by realSaddy** [comment]
I'll continue to work on these. As this is rather a large PR, I would appreciate any help with testing and the C++ implementation. I've updated the OP to include a comment with the planned strategies.

**Comment by mcm001** [comment]
Looking awesome! Hopefully we (my old team) will be able to get some good pose data out of our competition this weekend to feed into unit tests

**Comment by realSaddy** [comment]
> Looking awesome! Hopefully we (my old team) will be able to get some good pose data out of our competition this weekend to feed into unit tests

That'd be great! I added some dummy data in the tests in the meantime. 

The base java implementation is done. C++ isn't my forte, but I'll start working on adding a C++ translation. Thoughts on merging this PR and making a separate PR for the C++, or just getting everything done in this? I'm fine with either.

File: photon-lib/src/main/java/org/photonvision/RobotPoseEstimator.java [hunk 1]
```diff
@@ -257,7 +257,16 @@ private Pair<Pose3d, Double> averageBestTargetsStrategy() {
...
+                try {
+                    totalAmbiguity += 1. / target.getPoseAmbiguity();
+                } catch (ArithmeticException e) {
+                    DriverStation.reportWarning(
+                            "[RobotPoseEstimator] A total ambiguity of zero exists, using that pose instead!",
+                            false);
+                    return Pair.of(
+                            targetPose.transformBy(target.getBestCameraToTarget().inverse()),
+                            p.getFirst().getLatestResult().getLatencyMillis());
+                }
```
Comment (mcm001) [inline photon-lib/src/main/java/org/photonvision/RobotPoseEstimator.java:269]: this try catch is slightly curst lol

Comment (realSaddy) [reply photon-lib/src/main/java/org/photonvision/RobotPoseEstimator.java:269]: Yeah; we want to do an inversed weighted average (lower ambiguity = higher weight), so the totalAmbiguity needs to be the summation of the inverse ambiguities; reduces time complexity in the 0 edge case 😉  .

**Comment by realSaddy** [comment]
C++ written, still need to write tests. I would greatly appreciate any help on that. For some reason, my C++ tests aren't running... strange.

Opened as "ready for review" to get feedback.

**Comment by mdurrani808** [comment]
Not qualified to review, but any chances the following could be added to this PR:
`    /**
     * Estimates the pose of the robot in the field coordinate system, given the id of the fiducial, the robot relative to the
     * camera, and the target relative to the camera.
     * @param id id of the AprilTag detected, returned by PhotonVision
     * @param robotToCamera Transform3D of the robot relative to the camera. Origin of the robot is defined as the center.
     * @param cameraToTarget Transform3D of the target relative to the camera, returned by PhotonVision
     * @return Robot position relative to the field. Returns null if target id is not found with target map.
     */
    public Transform3d getFieldToRobot(Transform3d tagPose, Transform3d robotToCamera, Transform3d cameraToTarget) {
        try {
            return tagPose.plus(cameraToTarget.inverse()).plus(robotToCamera.inverse());
        } catch (Exception e) {
            return null;
        }
    }`

Or some variation (ton on if the id should be accepted as a parameter or the pose). If not, I can PR it later, but having it all in one would make it so we can get the examples written along with relevant documentation.

**Comment by realSaddy** [comment]
>Not qualified to review, but any chances the following could be added to this PR:

@mdurrani808 I think that method would best be added to PhotonUtils; this class is a more "managed" version. If you concur, I think it would be best to have that as a separate PR.

**Comment by mdurrani808** [comment]
> @mdurrani808 I think that method would best be added to PhotonUtils; this class is a more "managed" version. If you concur, I think it would be best to have that as a separate PR.

Makes sense, sounds good.

**Comment by realSaddy** [comment]
Everything is now done and ready for final review. Note that CI / photonserver-build-offline-docs is failing until #596 is merged.
LMK if any changes are requested!

**Review by gerth2 — CHANGES_REQUESTED** [review]

File: photon-lib/src/main/java/org/photonvision/RobotPoseEstimator.java [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+            return lastPose;
+        }
+
+        // Pick the lowest and do the heavy calculations
+        PhotonTrackedTarget bestTarget = cameras.get(lowestAI).getFirst().getLatestResult().targets.get(lowestAJ);
+
+        // If the map doesn't contain the ID fail
+        if (!aprilTags.containsKey(bestTarget.getFiducialId())) {
+            DriverStation.reportError(
+                    "[RobotPoseEstimator] Tried to get pose of unknown April Tag: " + bestTarget.getFiducialId(),
```
[the remark this answers is not in the document]
Comment (realSaddy) [reply photon-lib/src/main/java/org/photonvision/RobotPoseEstimator.java]: Possibly...
The reason for putting it as I foresaw teams trying to debug why it wasn't working when the issue is they are using a tag that isn't defined.

Thoughts on changing it so it will only print once per tag? Can be accomplished with a hashset.

File: photon-lib/src/main/java/org/photonvision/RobotPoseEstimator.java [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+     */
+    public RobotPoseEstimator(Map<Integer, Pose3d> aprilTags, PoseStrategy strategy, ArrayList<Pair<PhotonCamera, Transform3d>> cameras) {
+        this.aprilTags = aprilTags;
+        this.strategy = strategy;
+        this.cameras = cameras;
+    }
+
+
+    public Pose3d update() {
+        if (cameras.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (realSaddy) [reply photon-lib/src/main/java/org/photonvision/RobotPoseEstimator.java:94]: If so I'd make it toggleable. I'm not well versed in using Field2D beyond the simple x,y. This would also require the map to be specified, right? Decreases game generality.

File: photon-lib/src/main/native/cpp/photonlib/RobotPoseEstimator.cpp [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+                      "");
+  }
+  return std::make_pair(lastPose, units::millisecond_t(0));
+}
+
+std::pair<frc::Pose3d, units::millisecond_t>
+RobotPoseEstimator::LowestAmbiguityStrategy() {
+  int lowestAI = -1;
+  int lowestAJ = -1;
+  double lowestAmbiguityScore = 10;
```
[the remark this answers is not in the document]
Comment (realSaddy) [reply photon-lib/src/main/native/cpp/photonlib/RobotPoseEstimator.cpp:93]: Sure I'll switch this. I initially put 10 as I figured if you have an ambiguity above 10 (which shouldn't be possibly -- fact check me?) then you probably don't want to even use that pose

File: photon-lib/src/main/native/cpp/photonlib/PhotonCamera.cpp [hunk 5]
```diff
@@ -53,6 +53,7 @@ PhotonCamera::PhotonCamera(const std::string& cameraName)
                    cameraName) {}
 
 PhotonPipelineResult PhotonCamera::GetLatestResult() {
+  if (test) return testResult;
```
[the remark this answers is not in the document]
Comment (realSaddy) [reply photon-lib/src/main/native/cpp/photonlib/PhotonCamera.cpp:56]: Another way is to make the method virtual and overwrite it in an injector class like I did in the Java implementation. I tried doing that but ran into a few hurdles and decided to just take the simple way

File: photon-lib/src/main/java/org/photonvision/RobotPoseEstimator.java [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+                aprilTags
+                        .get(bestTarget.getFiducialId())
+                        .transformBy(bestTarget.getBestCameraToTarget().inverse())
+                        .transformBy(cameras.get(lowestAI).getSecond().inverse()),
+                cameras.get(lowestAI).getFirst().getLatestResult().getLatencyMillis());
+    }
+
+    private Pair<Pose3d, Double> closestToCameraHeightStrategy() {
+        // Loop over each ambiguity of all the cameras
+        double smallestHeightDifference = 10e9;
```
[the remark this answers is not in the document]
Comment (realSaddy) [reply photon-lib/src/main/java/org/photonvision/RobotPoseEstimator.java:172]: 👍

File: photon-lib/src/main/java/org/photonvision/RobotPoseEstimator.java [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+                if (!aprilTags.containsKey(target.getFiducialId())) {
+                    DriverStation.reportWarning(
+                            "[RobotPoseEstimator] Tried to get pose of unknown April Tag: "
+                                    + target.getFiducialId(),
+                            false);
+                    continue;
+                }
+                Pose3d targetPose = aprilTags.get(target.getFiducialId());
+                try {
+                    totalAmbiguity += 1. / target.getPoseAmbiguity();
```
[the remark this answers is not in the document]
Comment (realSaddy) [reply photon-lib/src/main/java/org/photonvision/RobotPoseEstimator.java:282]: Less ambiguous poses need to be weighted more. 

X_n * (1 / w_n) / (\sigma 1/w_n)

Accomplishes that while keeping it summing to one

File: photon-lib/src/main/java/org/photonvision/RobotPoseEstimator.java [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+                } catch (ArithmeticException e) {
+                    DriverStation.reportWarning(
+                            "[RobotPoseEstimator] A total ambiguity of zero exists, using that pose instead!",
```
[the remark this answers is not in the document]
Comment (realSaddy) [reply photon-lib/src/main/java/org/photonvision/RobotPoseEstimator.java:285]: It's not really an error — more of a debugging message. I'll switch it to be a comment

File: photon-lib/src/main/native/cpp/photonlib/RobotPoseEstimator.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (calcmogul) [reply photon-lib/src/main/native/cpp/photonlib/RobotPoseEstimator.cpp:93]: Use `std::numeric_limits<double>::infinity()` from `<limits>` instead of DBL_MAX.

File: photon-lib/src/main/java/org/photonvision/RobotPoseEstimator.java [hunk 9]
```diff
@@ -0,0 +1,370 @@
...
+        Translation3d transform = new Translation3d();
+        Rotation3d rotation = new Rotation3d();
+        double latency = 0;
+
+        for (Pair<Pose3d, Pair<Double, Double>> pair : tempPoses) {
+            try {
+                double weight = (1. / pair.getSecond().getFirst()) / totalAmbiguity;
+                transform = transform.plus(pair.getFirst().getTranslation().times(weight));
+                rotation = rotation.plus(pair.getFirst().getRotation().times(weight));
+                latency += pair.getSecond().getSecond() * weight; // NOTE: Average latency may not work well
```
[the remark this answers is not in the document]
Comment (calcmogul) [reply photon-lib/src/main/java/org/photonvision/RobotPoseEstimator.java:308]: This seems suspect, and no mathematical justification was given in the code. Where are these poses even from?

[the remark this answers is not in the document]
Comment (realSaddy) [reply photon-lib/src/main/java/org/photonvision/RobotPoseEstimator.java:308]: @calcmogul 
Less ambiguous poses need to be weighted more. Poses are from Photon's detection

X_n * (1 / w_n) / (\sigma 1/w_n)
Proof: 
(\sigma 1 / w_n) / (\sigma 1/w_n) = 1

[the remark this answers is not in the document]
Comment (calcmogul) [reply photon-lib/src/main/java/org/photonvision/RobotPoseEstimator.java:308]: But why average them all together? The correct thing to do is give them individually to the pose estimator with different measurement standard deviations. Averaging them into a single pose throws away information.

[the remark this answers is not in the document]
Comment (calcmogul) [reply photon-lib/src/main/java/org/photonvision/RobotPoseEstimator.java:308]: The other problem though is that averaging ambiguous poses is necessarily going to pull your state estimate away from the true value when one of the poses is correct and the other is completely incorrect. The ambiguity metric isn't an accurate reflection of the true probability distribution.

[the remark this answers is not in the document]
Comment (realSaddy) [reply photon-lib/src/main/java/org/photonvision/RobotPoseEstimator.java:308]: Yep, you're 100% right; I went in assuming that the ambiguity would be a true (gaussian) probability distribution -- even then, the latency wouldn't be mathematically correct. 
Any ideas on how to change it? I'm not even sure if it makes since to have this strategy in the first place...

[the remark this answers is not in the document]
Comment (calcmogul) [reply photon-lib/src/main/java/org/photonvision/RobotPoseEstimator.java:308]: > the latency wouldn't be mathematically correct

If the latencies are different, you have to feed them individually to the pose estimator so it can match them up with prior pose estimates before fusing them.

> Any ideas on how to change it? I'm not even sure if it makes since to have this strategy in the first place...

Particle filters and factor graphs can accurately represent multimodal distributions. Those are a lot more work to implement though. I would suggest dropping this strategy since the others handle the distribution a bit better (i.e., pick one target or the other based on some metric, but nothing in between).

[the remark this answers is not in the document]
Comment (JoshuaAN) [reply photon-lib/src/main/java/org/photonvision/RobotPoseEstimator.java:308]: It's possible to resolve ambiguity if multiple tags are in view by feeding all visible corners to SolvePNP (instead of each tag separately, then fusing each estimate in). This will converge to the true pose.

[the remark this answers is not in the document]
Comment (calcmogul) [reply photon-lib/src/main/java/org/photonvision/RobotPoseEstimator.java:308]: ^ I think libapriltag uses some kind of homography thing instead of PnP, but I like that idea.

File: photon-lib/src/main/java/org/photonvision/RobotPoseEstimator.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (mdurrani808) [reply photon-lib/src/main/java/org/photonvision/RobotPoseEstimator.java]: Once per tag sounds good tome.

File: photon-lib/src/main/java/org/photonvision/RobotPoseEstimator.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (gerth2) [reply photon-lib/src/main/java/org/photonvision/RobotPoseEstimator.java:94]: It can accept many poses. The background is actually set on the user side, it just shows shapes at the x/y/theta coordinates that are specified.

[the remark this answers is not in the document]
Comment (gerth2) [reply photon-lib/src/main/java/org/photonvision/RobotPoseEstimator.java:94]: Re: toggleable.... possibly? That would follow pattern with wpilib's sendable, and reduce NT server load if folks aren't using it. However, it's still on the user to add whatever is published to some dashboard to view, so it shouldn't be "in the way" even if they turn it on.

File: photon-lib/src/main/java/org/photonvision/RobotPoseEstimator.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (gerth2) [reply photon-lib/src/main/java/org/photonvision/RobotPoseEstimator.java:282]: I gotcha. I presumed that `totalAmbiguity` would by necessity be `sum of all ambiguities`. I buy the overall math though.

**Comment by realSaddy** [comment]
https://github.com/PhotonVision/photonvision/commit/bb63af601d3c88ca1a525c093d277413bb14c473#diff-060f4e35b6b11bfd815bca4146e4d956020dd01ad50c754922394f368d01a24c broke the CPP build. Fixed one issue (span), the other one is due to sourcing the geometry includes from the gradle repo. They aren't playing nicely when put into a pair  (see CI output).

**Comment by mcm001** [comment]
Taking a look at C++ now

File: photon-lib/src/main/native/cpp/photonlib/PhotonTrackedTarget.cpp [hunk 10]
```diff
@@ -34,14 +34,17 @@ namespace photonlib {
 
 PhotonTrackedTarget::PhotonTrackedTarget(
     double yaw, double pitch, double area, double skew, int id,
-    const frc::Transform3d& pose,
```
[the remark this answers is not in the document]
Comment (mcm001) [reply photon-lib/src/main/native/cpp/photonlib/PhotonTrackedTarget.cpp:37]: Yeah, old legacy reflective solvepnp targets will only have one pose. We can add a second ctor

[the remark this answers is not in the document]
Comment (mcm001) [reply photon-lib/src/main/native/cpp/photonlib/PhotonTrackedTarget.cpp:37]: Actually this doesn't super help us, does it? All the references to this constructor are internal to the library anyways

**Review by mcm001 — APPROVED** [review]
